src/System.ServiceModel.NetNamedPipe/src/System/ServiceModel/Channels/PipeConnection.cs

```diff
@@ -0,0 +1,576 @@
+        private bool _isShutdownWritten;
+
+        // timeout support
+        private string _timeoutErrorString;
+        private TransferOperation _timeoutErrorTransferOperation;
+
+        public PipeConnection(NamedPipeClientStream namedPipeClient, int connectionBufferSize)
+        {
+            _pipe = namedPipeClient ?? throw DiagnosticUtility.ExceptionUtility.ThrowHelperArgumentNull(nameof(namedPipeClient));
+            if (namedPipeClient == null)
+                throw DiagnosticUtility.ExceptionUtility.ThrowHelperArgumentNull(nameof(namedPipeClient));
+            if (!namedPipeClient.IsConnected)
+                throw DiagnosticUtility.ExceptionUtility.ThrowHelperArgumentNull(nameof(namedPipeClient));
+
+            _pipe = namedPipeClient;
+            _closeState = CloseState.Open;
+            _exceptionEventType = TraceEventType.Error;
+            _connectionBufferSize = connectionBufferSize;
+            _atEOFEvent = new ManualResetEventSlim(false);
+        }
+
+        public int ConnectionBufferSize
+        {
+            get
+            {
+                return _connectionBufferSize;
+            }
+        }
+
+        private static byte[] ZeroBuffer
+        {
+            get
+            {
+                return s_zeroBuffer;
+            }
+        }
+
+        public TraceEventType ExceptionEventType
+        {
+            get { return _exceptionEventType; }
+            set { _exceptionEventType = value; }
+        }
+
+        public void Abort()
+        {
+            Abort(null, TransferOperation.Undefined);
+        }
+
+        private void Abort(string timeoutErrorString, TransferOperation transferOperation)
+        {
+            CloseHandle(true, timeoutErrorString, transferOperation);
+        }
+
+        private Exception ConvertPipeException(PipeException pipeException, TransferOperation transferOperation)
+        {
+            return ConvertPipeException(pipeException.Message, pipeException, transferOperation);
+        }
+
+        private Exception ConvertPipeException(string exceptionMessage, PipeException pipeException, TransferOperation transferOperation)
+        {
+            if (_timeoutErrorString != null)
+            {
+                if (transferOperation == _timeoutErrorTransferOperation)
+                {
+                    return new TimeoutException(_timeoutErrorString, pipeException);
+                }
+                else
+                {
+                    return new CommunicationException(_timeoutErrorString, pipeException);
+                }
+            }
+            else if (_aborted)
+            {
+                return new CommunicationObjectAbortedException(exceptionMessage, pipeException);
+            }
+            else
+            {
+                return new CommunicationException(exceptionMessage, pipeException);
+            }
+        }
+
+        public async ValueTask<int> ReadAsync(Memory<byte> buffer, TimeSpan timeout)
```

Comment:
+1 to `CancellationToken` instead of `TimeSpan timeout`.

Reply:
We use IOThreads for running our own timer queue baseed on a Fibonacci priority queue as we had massive performance problems with the runtime's TimerManager (global lock around cancelling timers which CancellationToken does). Some of our perf tests were spending 30%+ CPU time registering and cancelling tiemrs. This was code copied from NetFx. I could rework that to create our own equivalent of a CancellationTokenSource which uses our own timer mechanism under the hood. We do have a problem of when a CancellationToken gets cancelled that we need to throw an exception with the original Timeout and that's not easily retrieved from a CancellationToken. I'll give this some thought and try to come up with a solution.